File: jl/BeginnersSelection/04_Coins.jl
# https://atcoder.jp/contests/abs/tasks/abc087_b

(
  (readline(), readline(), readline(), readline())
  |> a -> parse.(Int, a)
  |> t -> (
    (a, b, c, x) -> (
      (0 : min(a, floor(x / 500)))
      |> is -> map(i -> (
        (0 : min(b, floor((x - 500 * i) / 100)))
        |> js -> filter(j -> (
          x - 500 * i - 100 * j <= 50 * c
        ), js)
        |> length
      ), is)
      |> sum
    )
  )(t[1], t[2], t[3], t[4])
  |> println
)

# https://atcoder.jp/contests/abs/submissions/22560065

File: rb/BeginnersSelection/04_Coins.rb
# https://atcoder.jp/contests/abs/tasks/abc087_b

class Object
  def let(f)
    f.call self
  end
end

[gets, gets, gets, gets]
  .map(&:to_i)
  .let(->((a, b, c, x)) {
    (0..[a, (x / 500).floor].min)
      .map { |i|
        (0..[b, ((x - 500 * i) / 100).floor].min)
          .filter { |j| x - 500 * i - 100 * j <= 50 * c }
          .size
      }
      .sum
  })
  .let(method(:puts))

# https://atcoder.jp/contests/abs/submissions/21519442
